packages/testsweets/lib/src/ui/shared/popup_menu/custom_popup_menu.dart

```diff
@@ -0,0 +1,447 @@
+import 'dart:io';
+import 'dart:math' as math;
+
+import 'package:flutter/material.dart';
+
+enum PressType {
+  longPress,
+  singleClick,
+}
+
+enum PreferredPosition {
+  top,
+  bottom,
+}
+
+class CustomPopupMenuController extends ChangeNotifier {
+  bool menuIsShowing = false;
+
+  void showMenu() {
+    menuIsShowing = true;
+    notifyListeners();
+  }
+
+  void hideMenu() {
+    if (menuIsShowing == true) {
+      menuIsShowing = false;
+      notifyListeners();
+    }
+  }
+
+  void toggleMenu() {
+    menuIsShowing = !menuIsShowing;
+    notifyListeners();
+  }
+}
+
+Rect _menuRect = Rect.zero;
+
+class CustomPopupMenu extends StatefulWidget {
+  CustomPopupMenu({
+    required this.child,
+    required this.menuBuilder,
+    required this.pressType,
+    this.controller,
+    this.arrowColor = const Color(0xFF4C4C4C),
+    this.showArrow = true,
+    this.barrierColor = Colors.black12,
+    this.arrowSize = 10.0,
+    this.horizontalMargin = 10.0,
+    this.verticalMargin = 10.0,
+    this.position,
+    this.menuOnChange,
+    required this.onLongPressMoveUpdate,
+    required this.onLongPressUpWhilePopupHidden,
+    required this.onLongPressDown,
+    this.enablePassEvent = true,
+  });
+
+  final Widget child;
+  final PressType pressType;
+  final bool showArrow;
+  final Color arrowColor;
+  final Color barrierColor;
+  final double horizontalMargin;
+  final double verticalMargin;
+  final double arrowSize;
+  final CustomPopupMenuController? controller;
+  final Widget Function() menuBuilder;
+  final PreferredPosition? position;
+  final void Function(bool)? menuOnChange;
+  final void Function(LongPressMoveUpdateDetails) onLongPressMoveUpdate;
+  final void Function() onLongPressUpWhilePopupHidden;
+  final void Function() onLongPressDown;
+
+  /// Pass tap event to the widgets below the mask.
+  /// It only works when [barrierColor] is transparent.
+  final bool enablePassEvent;
+
+  @override
+  _CustomPopupMenuState createState() => _CustomPopupMenuState();
+}
+
+class _CustomPopupMenuState extends State<CustomPopupMenu> {
+  RenderBox? _childBox;
+  RenderBox? _parentBox;
+  OverlayEntry? _overlayEntry;
+  CustomPopupMenuController? _controller;
+  bool _canResponse = true;
+
+  _showMenu() {
```

Comment:
All functions should have a return type to make long term maintenance easier.

Reply:
👍